nselib/idna.lua

```diff
@@ -0,0 +1,614 @@
+      -- To use this part of code, add disallowed_STD3_mapped and disallowed_STD3_valid
+      -- codepoints to idnaMappings.lua. For now, we ignore these because idnaMappings.lua
+      -- is set to support only for the latest version of IDNA.
+      if UseSTD3ASCIIRules then
+        if type(lookup) == "table" then
+          if lookup.status == "disallowed_STD3_valid" or lookup.status == "disallowed_STD3_mapped" then
+            table.insert(disallowedCodePoints, cp)
+          end
+        end
+      end
+    end
+  end
+
+  decoded_tbl = concat_table_in_tables(decoded_tbl)
+
+  -- If UseSTD3ASCIIRules=false, then disallowed_STD3_mapped values are considered
+  -- as mapped codepoints and are mapped with the input.
+  -- To use this part of code, add disallowed_STD3_mapped and disallowed_STD3_valid
+  -- codepoints to idnaMappings.lua. For now, we ignore these because idnaMappings.lua
+  -- is set to support only for the latest version of IDNA.
+  if not useSTD3ASCIIRules then
+    for index, cp in ipairs(decoded_tbl) do
+      local lookup = idnaMappings[cp]
+      if type(lookup) == "table" then
+        if lookup.status == "disallowed_STD3_mapped" then
+          decoded_tbl[index] = lookup[1]
+        end
+      end
+    end
+  end
+
+  decoded_tbl = concat_table_in_tables(decoded_tbl)
+
+  return decoded_tbl, disallowedCodePoints
+end
+
+
+-- Validate the input based on IDNA codepoints validation rules.
+--
+-- @param tableOfTables Table of codepoints of the splitted input.
+-- @param checkHyphens Boolean flag checks for 0x002D in unusual places.
+function validate(tableOfTables, checkHyphens)
+
+  if checkHyphens == nil then
+    checkHyphens = true
+  end
+
+  -- Validates the list of input codepoints.
+  for _, tbl in ipairs(tableOfTables) do
+
+    if checkHyphens then
+
+      -- Checks the 3rd and 4th position of input.
+      if (tbl[3] and tbl[3] == 0x002D) or (tbl[4] and tbl[4] == 0x002D) then
+        return false
+      end
+
+      -- Checks for starting and ending of input.
+      if tbl[1] == 0x002D or tbl[#tbl] == 0x002D then
+        return false
+      end
+
+    end
+
+    for _, v in ipairs(tbl) do
+      if v == 0x002E then
+        return false
+      end
+    end
+
+    -- TODO:
+    -- 1. Add validation for checkBidi, checkJoiners (if required).
+    -- 2. The label must not begin with a combining mark, that is: General_Category=Mark.
+  end
+
+  return true
+
+end
+
+-- This function converts the input codepoints into ASCII text based on IDNA rules.
+--
+-- @param codepoints Table of codepoints of decoded input.
+-- @param tbl Table of optional params.
+-- @param checkHyphens Boolean flag for checking hyphens presence in input.
+-- @param checkBidi Boolean flag to represent if the input is of Bidi type.
+-- @param checkJoiners Boolean flag to check for ContextJ rules in input.
+-- @param useSTD3ASCIIRules Boolean value to represent ASCII rules.
+-- @param transitionalProcessing Boolean value.
```

Comment:
Add default values to the NSEdoc and note any flags that are not supported. We can't rely on consumers to read the code; at best they will look at the documentation, so it must be correct.

Reply:
Done.